Optimizations

spullara opened the pull request:
Added a micro benchmark and then made some optimizations:

(Executed on a Macbook Pro 13", 2.7 ghz i7, -Xmx1G -Xms1G)
Before Average: 3.677476679731188
After Average: 4.712168415220144
~28% speed boost

The optimizations include (in effect order):
- Removed String.format and replaced it with StringBuilder
- Removed a + from the hash tag regex that wasn't necessary
- Changed from String.split to a custom implementation

The current split between the various stages are:
- autoLinkHashtags: 60% (the unicode regex ranges are more than 50% of the cost of this method)
- autoLinkURLs: 30%
- autoLinkUsernamesAndLists: 5%
- other: 5%

Attempted optimizations that were not optimizations:
- Change references to String to CharSequences and avoid the StringBuilder.toString() at the end of each link phase -- there must be internal conversions to String as this is _slower_

The next step would be to replace regexes with a custom parser. Finally, if you could combine all the stages in a reasonable way that would probably have a pretty big benefit.

Commit:
Add benchmark so we can measure progress

Commit:
update the split method and remove uses of String.format

Commit:
modify the benchmark to give more consistent results

Commit:
match some of the formatting that was already present

Commit:
match some of the formatting that was already present

Commit:
match some of the formatting that was already present

keitaf commented:
Thank you for the pull request! The optimization looks great.

A couple of comments:
- Can we use StringBuilder instead of StringBuffer? Java5 and Android Gingerbread supports StringBuilder, so I believe it's safe to switch.
- Since split() is only used in Autolink.java and the second parameter is always "<>", can we rewrite split() to simply split it by '<' or '>' ? I mean, something like the following.
  
  List<String> splitByBrackets(String text) {
    int index;
    List<String> chunks = new ArrayList<String>();
  
    while ((index = indexOfBrackets(text)) >= 0 ){
      chunks.add(text.substring(0, index));
      index++;
      text = text.substring(index, text.length());
    }
    chunks.add(text);
  }
  
  private static int indexOfBrackets(String text) {
    for (int i = 0 ; i < text.length(); i++) {
      char c = text.charAt(i);
      if (c == '<' || c == '>') {
        return i;
      }
    }
    return -1;
  }
- Can we not split ConformanceTest into 2? Can Benchmark simply use the original ConformanceTest (with some modification to make some variable non-private), or Can Benchmark be totally independent of ConformanceTest?

Commit:
unrefactor conformancetest

spullara commented:
StringBuffer - unfortunately the matcher only supports StringBuffer,
otherwise I would have used StringBuilder
Split - seems reasonable as long as we don't expect to use it for more cases
in the future
Benchmark - I updated the branch with the changes.

Sam

On Mon, Oct 10, 2011 at 2:16 PM, Keita Fujii <
reply@reply.github.com>wrote:

> Thank you for the pull request! The optimization looks great.
> 
> A couple of comments:
> - Can we use StringBuilder instead of StringBuffer? Java5 and Android
>   Gingerbread supports StringBuilder, so I believe it's safe to switch.
> - Since split() is only used in Autolink.java and the second parameter is
>   always "<>", can we rewrite split() to simply split it by '<' or '>' ? I
>   mean, something like the following.
>   
>   List<String> splitByBrackets(String text) {
>    int index;
>   List<String> chunks = new ArrayList<String>();
>   
>   while ((index = indexOfBrackets(text)) >= 0 ){
>    chunks.add(text.substring(0, index));
>    index++;
>    text = text.substring(index, text.length());
>   }
>   chunks.add(text);
>   }
>   
>   private static int indexOfBrackets(String text) {
>    for (int i = 0 ; i < text.length(); i++) {
>      char c = text.charAt(i);
>      if (c == '<' || c == '>') {
>        return i;
>      }
>   }
>   return -1;
>   }
> - Can we not split ConformanceTest into 2? Can Benchmark simply use the
>   original ConformanceTest (with some modification to make some variable
>   non-private), or Can Benchmark be totally independent of ConformanceTest?
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/twitter/twitter-text-java/pull/5#issuecomment-2352381

keitaf commented on src/com/twitter/Autolink.java:
```diff
@@ -83,68 +85,68 @@ public class Autolink {
         while (matcher.find()) {
           if (matcher.group(Regex.AUTO_LINK_USERNAME_OR_LISTS_GROUP_LIST) == null ||
-              matcher.group(Regex.AUTO_LINK_USERNAME_OR_LISTS_GROUP_LIST).equals("")) {
+                  matcher.group(Regex.AUTO_LINK_USERNAME_OR_LISTS_GROUP_LIST).equals("")) {
```
Can you replace equals("") with isEmpty()?

keitaf commented on src/com/twitter/Autolink.java:
```diff
@@ -337,4 +346,38 @@ public class Autolink {
+    int current = 0;
+    boolean onemore = false;
+    while (current < length || onemore) {
+      int start = current;
```
I guess we can remove this 'start' variable?
in line 360, s/start/current/
swap 370 and 372, then s/start/current/
move 378 before 374, then s/start/current/

keitaf commented on src/com/twitter/Autolink.java:
```diff
@@ -337,4 +346,38 @@ public class Autolink {
+    boolean onemore = false;
+    while (current < length || onemore) {
+      int start = current;
+      int i = -1;
```
It took me a while to figure out what this' i' is for.
Can we use more descriptive name, like 'minIndex'?

Also, can we rewrite this part like this?

int minIndex = -1;
for (char c: d.toCharArray()) {
 minIndex = Math.min(minIndex, s.indexOf(c, start));
}

spullara replied:
I like all the changes you've suggested. I have no problem with you changing the branch directly. Might be easier then me trying to reapply your suggestions.

Sam

On Oct 17, 2011, at 11:55 AM, Keita Fujii wrote:

> > @@ -337,4 +346,38 @@ public class Autolink {
> >   public void setNoFollow(boolean noFollow) {
> >     this.noFollow = noFollow;
> >   }
> > +
> > -  // The default String split is horribly inefficient
> > -  private static Iterable<String> split(final String s, final String d) {
> > -    List<String> strings = new ArrayList<String>();
> > -    int length = s.length();
> > -    int current = 0;
> > -    boolean onemore = false;
> > -    while (current < length || onemore) {
> > -      int start = current;
> > -      int i = -1;
> 
> It took me a while to figure out what this' i' is for.
> Can we use more descriptive name, like 'minIndex'?
> 
> Also, can we rewrite this part like this?
> 
> int minIndex = -1;
> for (char c: d.toCharArray()) {
> minIndex = Math.min(minIndex, s.indexOf(c, start));
> }
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/twitter/twitter-text-java/pull/5/files#r174386

keitaf replied:
OK then I'll make those changes directly. Thank you!!

keitaf commented:
Hi Sam,

Thank you for the changes. Overall looks good to me. I just made some minor comments, so can you please accommodate those? Then I'll merge this into the next release.

Thanks,
Keita

Commit:
accommodate suggestions on the pull request.

keitaf commented:
Hi Sam,

I made changes based on my suggestion. Can you please check if it looks ok?

Thanks,
Keita

spullara commented:
Looks good to me. Ship it!

spullara committed:
fix a bug in the way query strings are handled

spullara committed:
Merge branch 'optimizations' of github.com:twitter/twitter-text-java into optimizations
